File: 152.maximum-product-subarray.py
#
# @lc app=leetcode id=152 lang=python3
#
# [152] Maximum Product Subarray
#

from typing import *
# @lc code=start
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0

        max_array = [1]
        min_array = [-1]
        for index, num in enumerate(nums):
            if min_array[-1] == 0:
                max_array.append(num)
                min_array.append(num)
                continue
            if num <0:
                max_array.append(min_array[index-1] * num)
                min_array.append(max_array[index-1] * num)
            else:
                max_array.append(max_array[index-1] * num)
                min_array.append(min_array[index-1] * num)
                
        return max(max_array)
        
# @lc code=end

if __name__ == "__main__":
    print(Solution().maxProduct([3,-1,4]))

File: golang/152.maximum-product-subarray.go
/*
 * @lc app=leetcode id=152 lang=golang
 *
 * [152] Maximum Product Subarray
 */
package main

import "fmt"

// @lc code=start
func maxProduct(nums []int) int {
	length := len(nums)
	dpMax := make([]int, length)
	dpMin := make([]int, length)
	copy(dpMax, nums)
	copy(dpMin, nums)
	for i := 1; i < length; i++ {
		num := nums[i]
		if dpMin[i] > dpMin[i-1]*nums[i] {
			dpMin[i] = dpMin[i-1] * nums[i]
		}
		if dpMin[i] > dpMax[i-1]*num {
			dpMin[i] = dpMax[i-1] * num
		}

		if dpMax[i] < dpMax[i-1]*num {
			dpMax[i] = dpMax[i-1] * num
		}
		if dpMax[i] < dpMin[i-1]*nums[i] {
			dpMax[i] = dpMin[i-1] * nums[i]
		}
	}
	max := dpMax[0]
	for i := 0; i < len(dpMax); i++ {
		if max < dpMax[i] {
			max = dpMax[i]
		}
	}
	return max
	// return dpMax[len(dpMax)-1]
}

// @lc code=end

func main() {
	fmt.Println(maxProduct([]int{2, 3, -2, 4}))
	// fmt.Println(maxProduct([]int{-2, 0, -1}))
}
